File: antia_numerical_methods/f77/sub16/secani.f
!     Real zero of a given function using secant iteration
!     Function is calculated as FX*2**JF
!     This subroutine uses reverse communication to calculate function
!     values. If IER<0 the function should be evaluated and SECANI should
!     be called back with new function value. Calculation of function
!     value should not change any other variables in the call statement.
!
!     X0 : (input) Initial guess for the zero
!     XL : (input) Lower limit of interval where zero is expected
!     XU : (input) Upper limit of interval where zero is expected
!     X : (output) Value of x at which the function evaluation is required.
!		If IER=0 then it will contain the final value of zero computed
!		by the routine.
!     F : (input) Calculated value of the function at X.
!		If subroutine exits with IER<0, then the calling routine should
!		calculate the function value at X and call SECANI with this value
!		stored in F and JF. Other variables should not be changed.
!     JF : (input) The exponent of function value, the function value
!		should be F*2**JF
!     REPS : (input) Required relative accuracy
!     AEPS : (input) Required absolute accuracy
!     		The estimated error should be less than MAX(AEPS, REPS*ABS(X))
!     IER : (input/output) Error parameter, IER=0 implies successful execution
!		Before the first call IER should be set to zero
!		IER<0 implies that execution is not over and the subroutine needs
!			a new function evaluation at X. After calculating the
!			function value SECANI should be called back.
!     		IER=40 implies that function value is equal at two points
!     			and it is not possible to continue the iteration
!     		IER=402 implies XL>X0 or XU<X0, in which case no calculations are done
!     		IER=422 implies that iteration goes outside the specified limits
!     		IER=423 implies that iteration failed to converge to specified accuracy
!
!	Required routines : None (Function is calculated by the calling program)

      SUBROUTINE SECANI(X0,XL,XU,X,F,JF,REPS,AEPS,IER)
      IMPLICIT REAL*16(A-H,O-Z)
      PARAMETER(NIS=75)
      SAVE
 
      IFL=-IER
!     Jump to proper location and continue execution
      IF(IFL.EQ.1) GO TO 1500

!     Initial call to subroutine, start from beginning
      IF(XL.GT.X0.OR.XU.LT.X0) THEN
!     If X0 is outside the specified interval (XL,XU) then quit
        IER=402
        RETURN
      ENDIF
 
      X=X0
!     Select the increment for the next point X+DX
      DX=(XU-X0)/100.
      IF(ABS(DX).LT.5.*MAX(REPS*ABS(X),AEPS)) DX=(XL-X0)/5.
      IF(ABS(DX).GT.0.1Q0*MAX(ABS(X),100.*AEPS))
     1       DX=SIGN(0.1Q0*MAX(ABS(X),100.*AEPS),DX)
      F1=0.0
      JF1=0
      L=0
 
1000  L=L+1
      IER=-1
!     To evaluate the function at X
      RETURN

1500  DX1=DX
      F1=F1*2.Q0**(JF1-JF)
      IER=0
 
      IF(F1-F.EQ.0.0) THEN
        IF(F.EQ.0.0) RETURN
!     If F1=F and F.NE.0, then quit
        IER=40
        RETURN
      ENDIF
 
!     The secant iteration
      IF(L.GT.1) DX=DX1*F/(F1-F)
      X=X+DX
      F1=F
      JF1=JF
 
      IF(ABS(DX).LT.MAX(REPS*ABS(X),AEPS).AND.L.GT.2) RETURN
      IF(X.LT.XL.OR.X.GT.XU) THEN
!     If iteration goes outside the specified limits (XL,XU), then quit
        IER=422
        RETURN
      ENDIF
      IF(L.LT.NIS) GO TO 1000
 
!     The iteration fails to converge
      IER=423
      END

File: projects/antia_numerical_methods@Beliavsky/c/fun/secani.c
/*      Real zero of a given function using secant iteration
	Function is calculated as FX*2**JF
	This function uses reverse communication to calculate function
	values. If IER<0 the function should be evaluated and SECANI should
	be called back with new function value. Calculation of function
	value should not change any other variables in the call statement.

	X0 : (input) Initial guess for the zero
	XL : (input) Lower limit of interval where zero is expected
	XU : (input) Upper limit of interval where zero is expected
	X : (output) Value of x at which the function evaluation is required.
		If IER=0 then it will contain the final value of zero computed
		by the function.
	F : (input) Calculated value of the function at X.
		If function exits with IER<0, then the calling function should
		calculate the function value at X and call SECANI with this value
		stored in F and JF. Other variables should not be changed.
	JF : (input) The exponent of function value, the function value
		should be F*2**JF
	REPS : (input) Required relative accuracy
	AEPS : (input) Required absolute accuracy
     		The estimated error should be less than max(AEPS, REPS*fabs(X))
	IER : (input/output) Error parameter, IER=0 implies successful execution
		Before the first call IER should be set to zero
		IER<0 implies that execution is not over and the function needs
			a new function evaluation at X. After calculating the
			function value SECANI should be called back.
     		IER=40 implies that function value is equal at two points
     			and it is not possible to continue the iteration
     		IER=402 implies XL>X0 or XU<X0, in which case no calculations are done
     		IER=422 implies that iteration goes outside the specified limits
     		IER=423 implies that iteration failed to converge to specified accuracy


	Required functions : None (Function value is calculated by the calling program)
*/

#include <math.h>


void secani(double x0, double xl, double xu, double *x, double f, int jf,
	double reps,double aeps, int *ier)

{
	int nis=75;
	static int l,jf1;
	static double f1,r1,dx,dx1;

	if(*ier==-1) goto func;

	if(xl>x0 || xu<x0) {*ier=402; return;}

	*x=x0;
/*	Select the increment for the next point X+DX */
	dx=(xu-x0)/100.0;
	r1=reps*fabs(*x); if(aeps>r1) r1=aeps;
	if(fabs(dx)<5.0*r1) dx=(xl-x0)/5.;
	r1=fabs(*x); if(r1<100.0*aeps) r1=100.0*aeps;
	if(fabs(dx)>0.1*r1) {
		if(dx>=0.0) dx=r1;
		else dx=-r1;
	}
	f1=0.0; jf1=0; l=0;

loop:	l=l+1; *ier=-1;
/*	To evaluate the function at x */
		return;

func:	dx1=dx;
		f1=f1*pow(2.0, (double) (jf1-jf));

		if(f1-f == 0.0) {
			if(f == 0.0) {*ier = 0; return ;}
/*	If F1=F and F!=0, then quit */
			else {*ier=40; return;}
		}

/*	The secant iteration */
		if(l>1) dx=dx1*f/(f1-f);
		*x=(*x)+dx;
		f1=f; jf1=jf;

		r1=reps*fabs(*x); if(aeps>r1) r1=aeps;
		if(fabs(dx)<r1 && l>2) {*ier=0; return;}
		if(*x<xl || (*x)>xu) {*ier=422; return;}
	if(l<nis) goto loop;

	*ier=423; return;
}
